File: problems/树/513.找树左下角的值.cpp
#include <iostream>
#include <queue>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int findBottomLeftValue(TreeNode* root) {
    queue<TreeNode*> q;
    q.push(root);
    int leftmost;
    while (!q.empty()) {
        TreeNode* node = q.front(); q.pop();
        leftmost = node->val;
        if (node->right) q.push(node->right);
        if (node->left)  q.push(node->left);
    }
    return leftmost;
}
int main() {
    TreeNode* root = new TreeNode(2);
    root->left  = new TreeNode(1);
    root->right = new TreeNode(3);
    cout << findBottomLeftValue(root) << endl;  // 1
    TreeNode* root2 = new TreeNode(1);
    root2->left  = new TreeNode(2);
    root2->right = new TreeNode(3);
    root2->left->left  = new TreeNode(4);
    root2->right->left  = new TreeNode(6);
    root2->right->right = new TreeNode(7);
    root2->right->right->left = new TreeNode(9);
    cout << findBottomLeftValue(root2) << endl;  // 9
    cout << findBottomLeftValue(new TreeNode(1)) << endl;  // 1
    return 0;
}

File: problems/树/513.找树左下角的值.py
#!/usr/bin/env python3
"""
513. 找树左下角的值
"""
from typing import Optional
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def find_bottom_left_value(root: Optional[TreeNode]) -> int:
    queue = deque([root])
    leftmost = root.val
    while queue:
        node = queue.popleft()
        leftmost = node.val
        if node.right: queue.append(node.right)
        if node.left: queue.append(node.left)
    return leftmost

if __name__ == "__main__":
    root1 = TreeNode(2, TreeNode(1), TreeNode(3))
    print(find_bottom_left_value(root1))  # 1
    root2 = TreeNode(1, TreeNode(2, TreeNode(4), None), TreeNode(3, TreeNode(6), TreeNode(7, TreeNode(9), None)))
    print(find_bottom_left_value(root2))  # 9
    print(find_bottom_left_value(TreeNode(1)))  # 1
